### ISSUE TITLE
[BUG] IRC openssl 1.1

### ISSUE THREAD

**Issue by Renzix** [hlissner/doom-emacs#1862]
**Describe the issue**
IRC (doom module that uses circe) doesnt connect to freenode. I believe its because openSSL 1.1 is broken for emacs. See this issue https://github.com/jorgenschaefer/circe/issues/340

Following the thread there was a fix and it did work for me. Also the thread said it was unlikely that a fix will come in upstream because tls.el has been obsoleted. 
```lisp
(setq tls-end-of-info
      (concat
       "\\("
       ;; `openssl s_client' regexp.  See ssl/ssl_txt.c lines 219-220.
       ;; According to apps/s_client.c line 1515 `---' is always the last
       ;; line that is printed by s_client before the real data.
       "^    Verify return code: .+\n\\(\\|^    Extended master secret: .+\n\\)---\n\\|"
       ;; `gnutls' regexp. See src/cli.c lines 721-.
       "^- Simple Client Mode:\n"
       "\\(\n\\|"                           ; ignore blank lines
       ;; According to GnuTLS v2.1.5 src/cli.c lines 640-650 and 705-715
       ;; in `main' the handshake will start after this message.  If the
       ;; handshake fails, the programs will abort.
       "^\\*\\*\\* Starting TLS handshake\n\\)*"
       "\\)"))
```

Note nothing errors it just says `Package tls is depreicated`

*Messages*
```
Doom loaded 274 packages across 55 modules in 0.578s
Package cl is deprecated
Loading /home/genzix/.emacs.d/.local/cache/recentf...done
Quit
Debug mode on
Starting new Ispell process aspell with default dictionary...
Package tls is deprecated
Decrypting /home/genzix/.password-store/irc/freenode.net.gpg...done
```

My config https://git.renzix.com/Renzix/Dotfiles/src/branch/master/.doom.d/config.org
Mirror https://gitlab.com/renzix/Dotfiles/blob/master/.doom.d/config.org

**Steps to reproduce**
1. Install DOOM emacs
2. Enable IRC
3. Setup IRC config
4. Make sure you are on openssl 1.1 (I think)

**System information**
<details><pre>
```
emacs   version    27.0.50
        features   XPM JPEG TIFF GIF PNG RSVG CAIRO IMAGEMAGICK SOUND GPM DBUS GLIB NOTIFY INOTIFY ACL GNUTLS LIBXML2 FREETYPE ZLIB GTK3 X11 XDBE XIM MODULES THREADS JSON PDUMPER LCMS2 GMP
        build      Oct 04, 2019
        buildopts  (--prefix=/usr --build=x86_64-pc-linux-gnu --host=x86_64-pc-linux-gnu --mandir=/usr/share/man --infodir=/usr/share/info --datadir=/usr/share --sysconfdir=/etc --localstatedir=/var/lib --disable-silent-rules --docdir=/usr/share/doc/emacs-vcs-27.0.9999 --htmldir=/usr/share/doc/emacs-vcs-27.0.9999/html --libdir=/usr/lib64 --program-suffix=-emacs-27-vcs --includedir=/usr/include/emacs-27-vcs --infodir=/usr/share/info/emacs-27-vcs --localstatedir=/var --enable-locallisppath=/etc/emacs:/usr/share/emacs/site-lisp --without-compress-install --without-hesiod --without-pop --with-file-notification=inotify --enable-acl --with-dbus --with-modules --without-gameuser --with-libgmp --with-gpm --with-json --without-kerberos --without-kerberos5 --with-lcms2 --with-xml2 --without-mailutils --without-selinux --with-gnutls --without-libsystemd --with-threads --without-wide-int --with-zlib --with-sound=alsa --with-x --without-ns --without-gconf --without-gsettings --without-toolkit-scroll-bars --with-gif --with-jpeg --with-png --with-rsvg --with-tiff --with-xpm --with-imagemagick --with-xft --with-cairo --without-harfbuzz --without-libotf --without-m17n-flt --with-x-toolkit=gtk3 --without-xwidgets 'CFLAGS=-march=native -O2 -pipe' CPPFLAGS= 'LDFLAGS=-Wl,-O1 -Wl,--as-needed')
        windowsys  x
        daemonp    server-running
doom    version    2.0.9
        build      HEAD -> develop, origin/develop, origin/HEAD fac6eabd7 2019-10-04 17:29:02 -0400
system  type       gnu/linux
        config     x86_64-pc-linux-gnu
        shell      /bin/bash
        uname      Linux 5.2.14 #1 SMP Sun Sep 15 23:50:25 EDT 2019 x86_64
        path       (/opt/powershell/6.2.0 ~/.local/bin ~/.perl6/bin ~/.cargo/bin ~/.go/bin ~/Dotfiles/Bin /usr/lib/llvm/9/bin /usr/lib/llvm/8/bin /usr/local/sbin /usr/local/bin /usr/sbin /usr/bin /sbin /bin /opt/bin /opt/cuda/bin ~/.fzf/bin /usr/libexec/emacs/27.0.50/x86_64-pc-linux-gnu/)
config  envfile    envvar-file
        elc-files  2
        modules    (:completion (company +auto) (helm +fuzzy) :ui doom hl-todo indent-guides modeline nav-flash ophints (popup +all +defaults) pretty-code treemacs vc-gutter vi-tilde-fringe window-select workspaces :editor evil file-templates fold multiple-cursors rotate-text snippets :emacs dired electric vc :tools eval flycheck flyspell (lookup +docsets) lsp magit :term eshell vterm :lang (cc +lsp +irony) common-lisp data emacs-lisp (go +lsp) (haskell +intero +lsp) (java +meghanada +eclim +lsp) latex (lua +lsp) markdown nix (org +attach +babel +capture +export +habit +present +protocol) perl (python +lsp) (rust +lsp) (scala +lsp +eclim) (sh +bash) web :app irc (write +write-mode +wordnut +langtool) :tools pass :config literate (default +bindings +smartparens))
        packages   (n/a)
        elpa       (n/a)
```
</pre></details>

**Comment by hlissner** [comment]
Would it make sense to only apply these settings in circe-mode buffers? Just to limit the scope of this hack, in case it has side-effects. e.g.

```elisp
  (setq-hook! 'circe-mode-hook
    tls-end-of-info
    (concat "\\("
            "^    Verify return code: .+\n\\(\\|^    Extended master secret: .+\n\\)---\n\\|"
            "^- Simple Client Mode:\n"
            "\\(\n\\|"
            "^\\*\\*\\* Starting TLS handshake\n\\)*"
            "\\)"))
```

Or perhaps these are better suited to `circe-channel-mode-hook`? I don't use circe so I wouldn't know.

**Comment by Renzix** [comment]
Testing it further on my desktop (gentoo) it doesnt work consistantly. Testing it on my laptop (arch) it works without the setq perfectly fine. The annoying thing is that they are the exact same version so I guess its either a patch from gentoo that messed with it as I even enabled all the compile time options I could. Sorry for wasting your time

**Comment by hperrey** [comment]
For me, it didn't quite work consistently either (on OpenSSL 1.1.1a, Ubuntu Mate 19.10 and Emacs 26.3). However, making a minor adjustment to the end-of-info identification seem to work fine so far:
``` elisp
(setq tls-end-of-info
      (concat
       "\\("
       ;; `openssl s_client' regexp.  See ssl/ssl_txt.c lines 219-220.
       ;; According to apps/s_client.c line 1515 `---' is always the last
       ;; line that is printed by s_client before the real data.
       "^    Verify return code: .+\n\\(\\|^    Extended master secret: .+\n\\)\\(\\|^    Max Early Data: .+\n\\)---\n\\|"
       ;; `gnutls' regexp. See src/cli.c lines 721-.
       "^- Simple Client Mode:\n"
       "\\(\n\\|"                           ; ignore blank lines
       ;; According to GnuTLS v2.1.5 src/cli.c lines 640-650 and 705-715
       ;; in `main' the handshake will start after this message.  If the
       ;; handshake fails, the programs will abort.
       "^\\*\\*\\* Starting TLS handshake\n\\)*"
       "\\)"))
```
If this could be incorporated into doom that would be great -- not sure where else it could be made available, save for tls.el. Let me know, I'd be happy to test any code changes!

**Comment by hlissner** [comment]
This fix was incorporated into Doom in #2727 thanks to @hperrey. Thanks for bringing it to my attention!
